libshpool/src/protocol.rs

```diff
@@ -120,19 +121,70 @@ impl<'data> ChunkExt<'data> for Chunk<'data> {
 }
 
 pub struct Client {
-    pub stream: UnixStream,
+    stream: UnixStream,
+}
+
+/// The result of creating a client, possibly with
+/// flagging some issues that need to be handled.
+pub enum ClientResult {
+    /// The created client, ready to go.
+    JustClient(Client),
+    /// There was a version mismatch between the client
+    /// process and the daemon process which ought to be
+    /// handled, though it is possible that some operations
+    /// will continue to work.
+    VersionMismatch {
+        /// A warning about a version mismatch that should be
+        /// displayed to the user.
+        warning: String,
+        /// The client, which may or may not work.
+        client: Client,
+    },
 }
 
 impl Client {
-    pub fn new<P: AsRef<Path>>(sock: P) -> anyhow::Result<Self> {
+    /// Create a new client
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new<P: AsRef<Path>>(sock: P) -> anyhow::Result<ClientResult> {
         let stream = UnixStream::connect(sock).context("connecting to shpool")?;
-        Ok(Client { stream })
-    }
 
-    pub fn write_connect_header(&mut self, header: ConnectHeader) -> anyhow::Result<()> {
-        let serialize_stream = self.stream.try_clone().context("cloning stream for reply")?;
-        encode_to(&header, serialize_stream).context("writing reply")?;
+        let daemon_version: VersionHeader = match decode_from(&stream) {
```

Comment:
It will break in a not-very-graceful way, which is currently what it does every time we touch the client to daemon protocol. I think what will happen is that the client will attempt to parse the version header as an response header (the exact type will depend on the operation the client is attempting), this will fail, and the client will exit with a failure code.

I can't really see how we can do any better because the current protocol has essentially no way to extend it with new functionality.

Reply:
makes sense. Let's hope this is the last silent breaking change, now that we have versioned protocols  :P